File: 2023/Leetcode/problem863.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildGraph(self, root):
        graph = {root.val: []}
        queue = deque([root])

        while len(queue) != 0:
            root = queue.popleft()

            if root.left:
                graph[root.val].append(root.left.val)
                graph[root.left.val] = [root.val]
                queue.append(root.left)

            if root.right:
                graph[root.val].append(root.right.val)
                graph[root.right.val] = [root.val]
                queue.append(root.right)

        return graph


    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        graph = self.buildGraph(root)

        result = []
        queue = deque([(target.val, 0)])
        visited = set()
        dist = 0

        while len(queue) != 0:
            curr, dist = queue.popleft()
            visited.add(curr)

            if dist == k:
                result.append(curr)
            else:
                for val in graph[curr]:
                    if val not in visited:
                        queue.append((val, dist + 1))

        return result

File: 2023/Leetcode/problem863.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public static HashMap<Integer, ArrayList<Integer>> buildGraph(TreeNode root) {
        HashMap<Integer, ArrayList<Integer>> graph = new HashMap<Integer, ArrayList<Integer>>();
        ArrayDeque<TreeNode> queue = new ArrayDeque<TreeNode>();

        queue.add(root);
        graph.put(root.val, new ArrayList<Integer>());

        while (queue.size() != 0) {
            root = queue.pop();

            if (root.left != null) {
                graph.get(root.val).add(root.left.val);
                graph.put(root.left.val, new ArrayList<Integer>());
                graph.get(root.left.val).add(root.val);

                queue.add(root.left);
            }
            
            if (root.right != null) {
                graph.get(root.val).add(root.right.val);
                graph.put(root.right.val, new ArrayList<Integer>());
                graph.get(root.right.val).add(root.val);

                queue.add(root.right);
            }
        }
        return graph;
    }

    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        HashMap<Integer, ArrayList<Integer>> graph = buildGraph(root);

        HashSet<Integer> visited = new HashSet<Integer>();
        ArrayList<Integer> result = new ArrayList<Integer>();
        ArrayDeque<Integer> queue = new ArrayDeque<Integer>();

        int curr;
        int distance;

        queue.add(target.val);
        queue.add(0);
        visited.add(target.val);

        while (queue.size() != 0) {
            curr = queue.pop();
            distance = queue.pop();

            if (distance == k) {
                result.add(curr);
            } else {
                for (Integer val : graph.get(curr)) {
                    if (!visited.contains(val)) {
                        queue.add(val);
                        queue.add(distance + 1);
                        visited.add(val);
                    }
                }
            }
        }
        return result;
    }
}